File: output/GPT3.5-Turbo/xCodeEval_remove_4000/C++/Java/c5127f5691b4754aa4e976d93cb6a983.java
import java.util.*;
import java.io.*;

class Main {
    static class FastReader {
        BufferedReader br;
        StringTokenizer st;

        public FastReader() {
            br = new BufferedReader(new InputStreamReader(System.in));
        }

        String next() {
            while (st == null || !st.hasMoreElements()) {
                try {
                    st = new StringTokenizer(br.readLine());
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
            return st.nextToken();
        }

        int nextInt() {
            return Integer.parseInt(next());
        }

        long nextLong() {
            return Long.parseLong(next());
        }

        double nextDouble() {
            return Double.parseDouble(next());
        }

        String nextLine() {
            String str = "";
            try {
                str = br.readLine();
            } catch (IOException e) {
                e.printStackTrace();
            }
            return str;
        }
    }

    static class Pair {
        int first, second;

        public Pair(int first, int second) {
            this.first = first;
            this.second = second;
        }
    }

    static class Modular<T> {
        public static final int mod = 998244353;
        private T value;

        public Modular() {
            value = null;
        }

        public Modular(T value) {
            this.value = value;
        }

        public T getValue() {
            return value;
        }

        public void setValue(T value) {
            this.value = value;
        }

        public static int normalize(int x) {
            int v;
            if (-mod <= x && x < mod) {
                v = x;
            } else {
                v = x % mod;
            }
            if (v < 0) {
                v += mod;
            }
            return v;
        }

        public Modular<T> add(Modular<T> other) {
            int result = normalize((Integer)value + (Integer)other.getValue());
            return new Modular<>(result);
        }

        public Modular<T> subtract(Modular<T> other) {
            int result = normalize((Integer)value - (Integer)other.getValue());
            return new Modular<>(result);
        }

        public Modular<T> multiply(Modular<T> other) {
            int result = normalize((Integer)value * (Integer)other.getValue());
            return new Modular<>(result);
        }

        public Modular<T> divide(Modular<T> other) {
            int inverse = inverse((Integer)other.getValue(), mod);
            int result = normalize((Integer)value * inverse);
            return new Modular<>(result);
        }

        public int inverse(int a, int m) {
            int u = 0, v = 1;
            while (a != 0) {
                int t = m / a;
                m -= t * a;
                int temp = a;
                a = m;
                m = temp;
                u -= t * v;
                temp = u;
                u = v;
                v = temp;
            }
            assert(m == 1);
            return u;
        }
    }

    static int[] seg0, seg1;
    static Pair[] len;

    static boolean can(int n) {
        if (n == 1) {
            return true;
        }

        int x = n / 2, y = (n + 1) / 2;
        if (!can(x) || !can(y)) {
            return false;
        }
        int p0 = nxt[it][0], p1 = nxt[it][1];
        if (p0 == s.length() || ind[p0] + x - 1 >= o[0].size()) {
            p0 = s.length();
        } else {
            p0 = o[0].get(ind[p0] + x - 1);
        }
        if (p1 == s.length() || ind[p1] + y - 1 >= o[1].size()) {
            p1 = s.length();
        } else {
            p1 = o[1].get(ind[p1] + y - 1);
        }
        p0 = Math.min(p0, p1);
        if (p0 == s.length()) {
            return false;
        }
        it = p0 + 1;

        return true;
    }

    static int[] id, mas;

    static void gen(int l, int r) {
        if (l == r) {
            return;
        }
        int m = (l + r - 1) / 2;
        gen(l, m);
        gen(m + 1, r);
        int a = l, b = m + 1, c = l;
        while (a <= m && b <= r) {
            if (s.charAt(it) == '0') {
                mas[c++] = id[a];
                a++;
            } else {
                mas[c++] = id[b];
                b++;
            }
            it++;
        }
        while (a <= m) {
            mas[c++] = id[a];
            a++;
        }
        while (b <= r) {
            mas[c++] = id[b];
            b++;
        }
        for (int i = l; i <= r; i++) {
            id[i] = mas[i];
        }
    }

    static void go(int n, boolean cert) {
        it = 0;
        if (!cert) {
            if (can(n)) {
                System.out.println(n);
                for (int i = 0; i < n; i++) {
                    System.out.print(id[i] + 1 + " ");
                }
                System.out.println();
                System.exit(0);
            }
        } else {
            it = 0;
            can(n);
            assert(it == s.length());

            it = 0;
            id = new int[n];
            for (int i = 0; i < n; i++) {
                id[i] = i;
            }
            gen(0, n - 1);
            assert(it == s.length());
            int[] p = new int[n];
            for (int i = 0; i < n; i++) {
                p[id[i]] = i;
            }
            System.out.println(n);
            for (int i : p) {
                System.out.print(i + 1 + " ");
            }
            System.out.println();
            System.exit(0);
        }
    }

    static int[][] seg;
    static Pair[] len;
    static ArrayList<Integer>[] o;
    static int[] ind;
    static int it;
    static int[][] nxt;

    public static void main(String[] args) {
        FastReader sc = new FastReader();

        int ts = 1;

        for (int its = 1; its <= ts; its++) {
            int maxn = 100010;
            seg0 = new int[maxn];
            seg1 = new int[maxn];
            len = new Pair[maxn];
            for (int i = 2; i < maxn; i++) {
                int x = i / 2, y = i - x;
                len[i] = new Pair(len[x].first + len[y].first + x, len[x].second + len[y].second + x + y - 1);
                seg0[i] = seg0[x] + seg0[y];
                seg1[i] = seg1[x] + seg1[y];
                seg0[i] += x;
                seg1[i] += y;
            }
            s = sc.next();
            int[] counter = new int[2];
            nxt = new int[s.length() + 1][2];
            ind = new int[s.length()];
            o = new ArrayList[2];
            o[0] = new ArrayList<>();
            o[1] = new ArrayList<>();
            for (int i = 0; i < s.length(); i++) {
                int c = s.charAt(i) - '0';
                counter[c]++;
                ind[i] = o[c].size();
                o[c].add(i);
            }
            nxt[s.length()][0] = nxt[s.length()][1] = s.length();
            for (int i = s.length() - 1; i >= 0; i--) {
                for (int t = 0; t < 2; t++) {
                    nxt[i][t] = nxt[i + 1][t];
                }
                nxt[i][s.charAt(i) - '0'] = i;
            }
            int l = 0, r = 100010;
            while (r - l > 1) {
                int m = (l + r) / 2;
                if (can(m)) {
                    l = m;
                } else {
                    r = m;
                }
            }
            go(l, true);
        }
    }
}

File: output/GPT3.5-Turbo/xCodeEval_remove_4000/C++/C/c5127f5691b4754aa4e976d93cb6a983.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAXN 100010

typedef struct {
    int fr;
    int sc;
} Pair;

char s[MAXN];
int o[2][MAXN];
int ind[MAXN];
Pair nxt[MAXN];
int it;

typedef struct {
    int value;
} Modular;

Modular inverse(Modular a, Modular m) {
    Modular u = {0};
    Modular v = {1};
    while (a.value != 0) {
        Modular t = {m.value / a.value};
        m.value -= t.value * a.value;
        int temp = a.value;
        a.value = m.value;
        m.value = temp;
        u.value -= t.value * v.value;
        int temp2 = u.value;
        u.value = v.value;
        v.value = temp2;
    }
    return u;
}

Modular Modular_add(Modular lhs, Modular rhs, Modular mod) {
    Modular result = {lhs.value + rhs.value};
    if (result.value >= mod.value) {
        result.value -= mod.value;
    }
    return result;
}

Modular Modular_sub(Modular lhs, Modular rhs, Modular mod) {
    Modular result = {lhs.value - rhs.value};
    if (result.value < 0) {
        result.value += mod.value;
    }
    return result;
}

Modular Modular_mul(Modular lhs, Modular rhs, Modular mod) {
    Modular result = {lhs.value * rhs.value % mod.value};
    return result;
}

Modular Modular_div(Modular lhs, Modular rhs, Modular mod) {
    Modular inv = inverse(rhs, mod);
    return Modular_mul(lhs, inv, mod);
}

Modular power(Modular a, int b, Modular mod) {
    Modular x = a;
    Modular res = {1};
    int p = b;
    while (p > 0) {
        if (p & 1) {
            res = Modular_mul(res, x, mod);
        }
        x = Modular_mul(x, x, mod);
        p >>= 1;
    }
    return res;
}

Modular normalize(int x, Modular mod) {
    int v;
    if (-mod.value <= x && x < mod.value) {
        v = x;
    } else {
        v = x % mod.value;
    }
    if (v < 0) {
        v += mod.value;
    }
    Modular result = {v};
    return result;
}

Modular mod = {998244353};

int can(int n) {
    if (n == 1) {
        return 1;
    }
    
    int IT = it;
    int x = n / 2, y = (n + 1) / 2;
    if (!can(x) || !can(y)) {
        return 0;
    }
    int p0 = nxt[it].fr, p1 = nxt[it].sc;
    if (p0 == strlen(s) || ind[p0] + x - 1 >= o[0][0]) {
        p0 = strlen(s);
    } else {
        p0 = o[0][ind[p0] + x - 1];
    }
    if (p1 == strlen(s) || ind[p1] + y - 1 >= o[1][0]) {
        p1 = strlen(s);
    } else {
        p1 = o[1][ind[p1] + y - 1];
    }
    p0 = p0 < p1 ? p0 : p1;
    if (p0 == strlen(s)) {
        return 0;
    }
    it = p0 + 1;
    
    return 1;
}

void gen(int l, int r) {
    if (l == r) {
        return;
    }
    int m = (l + r - 1) / 2;
    gen(l, m);
    gen(m + 1, r);
    int a = l, b = m + 1, c = l;
    while (a <= m && b <= r) {
        if (s[it] == '0') {
            mas[c++] = id[a];
            a++;
        } else {
            mas[c++] = id[b];
            b++;
        }
        it++;
    }
    while (a <= m) {
        mas[c++] = id[a];
        a++;
    }
    while (b <= r) {
        mas[c++] = id[b];
        b++;
    }
    for (int i = l; i <= r; i++) {
        id[i] = mas[i];
    }
}

void go(int n, int cert) {
    it = 0;
    if (!cert) {
        if (!can(n)) {
            printf("0\n");
        } else {
            printf("1\n");
        }
        return;
    }
    it = 0;
    can(n);
    if (it == strlen(s)) {
        it = 0;
        for (int i = 0; i < n; i++) {
            id[i] = i;
        }
        gen(0, n - 1);
        for (int i = 0; i < n; i++) {
            printf("%d ", id[i] + 1);
        }
        printf("\n");
        exit(0);
    }
}

void solve() {
    for (int i = 2; i < MAXN; i++) {
        int x = i / 2, y = i - x;
        len[i].fr = len[x].fr + len[y].fr + x;
        len[i].sc = len[x].sc + len[y].sc + x + y - 1;
        seg[i][0] = seg[x][0] + seg[y][0];
        seg[i][1] = seg[x][1] + seg[y][1];
        seg[i][0] += x;
        seg[i][1] += y;
    }
    scanf("%s", s);
    int counter[2] = {0};
    int len_s = strlen(s);
    for (int i = 0; i < len_s; i++) {
        int c = s[i] - 48;
        counter[c]++;
        ind[i] = o[c][0];
        o[c][0]++;
    }
    nxt[len_s].fr = nxt[len_s].sc = len_s;
    for (int i = len_s - 1; i >= 0; i--) {
        nxt[i].fr = nxt[i + 1].fr;
        nxt[i].sc = nxt[i + 1].sc;
        nxt[i].value = i;
    }
    int l = 0, r = 100010;
    while (r - l > 1) {
        int m = (l + r) / 2;
        go(m, 0);
        if (it == 0) {
            l = m;
        } else {
            r = m;
        }
    }
    go(l, 1);
}

int main() {
    freopen("../a.in", "r", stdin);
    freopen("../a.out", "w", stdout);

    int ts;
    ts = 1;
    
    for (int its = 1; its <= ts; its++) {
        solve();
    }

    return 0;
}
